### PR TITLE
Provenance and Advise Managers

### PR DESCRIPTION
**Pull request by KPostOffice**
## Description:
This provides two managers which interfaces with **Thoth** through the **Thamos** API library.

## This pull request includes:
- Code for two new managers
- **Adviser**
  - Automatically generate a pull request on successful execution of an advise job
  - Automatically open an issue if problems were found in the dependency stack
  - Issues and merge requests are such that only one can exist for the current head at a time. The current issue or merge request will always be the latest
- **Provenance**
  - Run provenance check on repository
  - Create a single issue which contains all errors that were found
  - Do nothing if no problems were found

### CODE REVIEW

File: kebechet/managers/thoth_provenance/thoth_provenance.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        self._cached_merge_requests = None
+        super().__init__(*args, **kwargs)
+
+    @property
+    def repo(self):
+        """Get repository on which we work on."""
+        return self._repo
+
+    @repo.setter
+    def repo(self, repo: git.Repo):
```
Comment (fridex) [inline kebechet/managers/thoth_provenance/thoth_provenance.py]: I see some of the methods/attributes used multiple times in these managers. Could you refactor them out in a way they share shared code? It will help maintain the codebase (e.g. introduce a base class with shared logic).

File: kebechet/managers/thoth_provenance/README.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
+Kebechet Update Manager
```
[the remark this answers is not in the document]
Comment (KPostOffice) [reply kebechet/managers/thoth_provenance/README.rst]: Realized I forgot to update the README, it should be fixed now

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+            f.write(json.dumps(lock_info))
+        return
+
+    def _issue_advise_error(self, adv_results: list, labels: list):
+        error_info = adv_results[0]["report"][0][0][0]
+        justification = error_info["justification"]
+        type_ = error_info["type"]
+        checksum = hashlib.md5(justification.encode("utf-8")).hexdigest()[:10]
+        self.sm.open_issue_if_not_exist(
+            f"{checksum}-{type_}: Automated kebechet thoth-advise Issue",
```
[the remark this answers is not in the document]
Comment (KPostOffice) [reply kebechet/managers/thoth_advise/thoth_advise.py]: It doesn't seem to bad since it only uses the first 10 characters, but I might be biased haha

File: kebechet/managers/thoth_provenance/thoth_provenance.py (hunk 1 above)
Comment (KPostOffice) [reply kebechet/managers/thoth_provenance/thoth_provenance.py]: I just placed them in the manager super class, does that work?

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+
+
+class ThothAdviseManager(ManagerBase):
+    """Manage updates of dependencies."""
+
+    def __init__(self, *args, **kwargs):
+        """Initialize update manager."""
+        self._repo = None
+        # We do API calls once for merge requests and we cache them for later use.
+        self._cached_merge_requests = None
```
[the remark this answers is not in the document]
Comment (KPostOffice) [reply kebechet/managers/thoth_advise/thoth_advise.py]: I'm trying to figure out how to implement event driven right now, I'll have to put some thought into what exactly needs to change

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 5]
```diff
@@ -128,14 +131,14 @@ def run(self, labels: list):
                     if res is not None:
                         break
                     LOGGER_.info(f"Advising failed, retrying ({i}/10)")
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: `_LOGGER`? Here and in other parts of sources.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        self, branch_name: str, labels: list, files: list
+    ) -> typing.Optional[int]:
+        """Open a pull/merge request for dependency update."""
+        commit_msg = "Auto generated update"
+        body = "Pipfile.lock updated by kebechet-thoth manager"
+
+        # Delete branch if it didn't change Pipfile.lock
+        diff = self.repo.git.diff("master", files)
+        if diff == "":
+            return
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: Probably some logged info will be appropriate.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+        return merge_request
+
+    @staticmethod
+    def _write_advise(adv_results: list):
+        lock_info = adv_results[0]["report"][0][1]["requirements_locked"]
+        with open("Pipfile.lock", "w+") as f:
+            LOGGER_.info('Writing to Pipfile.lock')
+            LOGGER_.debug(f"{json.dumps(lock_info)}")
+            f.write(json.dumps(lock_info))
+        return
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: No need to have `return` statement here.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+            branch_name = self._construct_branch_name()
+            branch = self.repo.git.checkout("-B", branch_name)
+            self._cached_merge_requests = self.sm.repository.merge_requests
+            if os.path.isfile("Pipfile"):
+                res = lib.advise_here()
+                for i in range(1, 11):
+                    if res is not None:
+                        break
+                    LOGGER_.info(f"Advising failed, retrying ({i}/10)")
+                    res = lib.advise_here(force=True)
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: Without `force` set to True. If requirements did not changed, we don't want to recompute results.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+                if res is None:
+                    LOGGER_.error("Advising failed")
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: Please log what went wrong here.

File: kebechet/managers/thoth_provenance/thoth_provenance.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+        """Run the provenance check bot."""
+        with cloned_repo(self.service_url, self.slug, depth=1) as repo:
+            self.repo = repo
+            if os.path.isfile("Pipfile") and os.path.isfile("Pipfile.lock"):
+                res = lib.provenance_check_here()
+                for i in range(1, 11):
+                    if res is not None:
+                        break
+                    LOGGER_.warning(f"Provenance check failed, retrying ({i}/10)")
+                    res = lib.provenance_check_here(force=True)
```
Comment (fridex) [inline kebechet/managers/thoth_provenance/thoth_provenance.py]: `force=False`

File: kebechet/managers/thoth_provenance/thoth_provenance.py [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+            labels=labels,
+        )
+
+    def run(self, labels: list):
+        """Run the provenance check bot."""
+        with cloned_repo(self.service_url, self.slug, depth=1) as repo:
+            self.repo = repo
+            if os.path.isfile("Pipfile") and os.path.isfile("Pipfile.lock"):
+                res = lib.provenance_check_here()
+                for i in range(1, 11):
```
Comment (fridex) [inline kebechet/managers/thoth_provenance/thoth_provenance.py]: I wonder - why is this needed?

Comment (KPostOffice) [reply kebechet/managers/thoth_provenance/thoth_provenance.py]: This is so that if the adviser failed, it will retry.  It has to force because otherwise it will repeatedly get the empty response.

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 8 above)
Comment (KPostOffice) [reply kebechet/managers/thoth_advise/thoth_advise.py]: The force is only if result is `None`, if a submission was made my first call does not force so it should get that result if it exists in the cache

Comment (fridex) [reply kebechet/managers/thoth_advise/thoth_advise.py]: I'm not sure if I follow here. Caching is done on the server side (backend should know what can be cached, for how long and how), the client should not be aware of it and dummy in this way.

File: kebechet/managers/thoth_provenance/thoth_provenance.py (hunk 11 above)
Comment (fridex) [reply kebechet/managers/thoth_provenance/thoth_provenance.py]: Kebechet should be simple here. If analysis failed, kebechet should report back it failed.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+    def _write_advise(adv_results: list):
+        lock_info = adv_results[0]["report"][0][1]["requirements_locked"]
+        with open("Pipfile.lock", "w+") as f:
+            _LOGGER.info('Writing to Pipfile.lock')
+            _LOGGER.debug(f"{json.dumps(lock_info)}")
+            f.write(json.dumps(lock_info))
+
+    def _issue_advise_error(self, adv_results: list, labels: list):
+        """Create an issue if advise fails."""
+        error_info = adv_results[0]["report"][0][0][0]
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: This array can hold multiple "reasonings". Can it be expanded in the issue body?

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+        )
+
+    def run(self, labels: list):
+        """Run Thoth Advising Bot."""
+        with cloned_repo(self.service_url, self.slug, depth=1) as repo:
+            self.repo = repo
+            branch_name = self._construct_branch_name()
+            branch = self.repo.git.checkout("-B", branch_name)
+            self._cached_merge_requests = self.sm.repository.merge_requests
+            if os.path.isfile("Pipfile"):
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: What happens if there is no `Pipfile`?

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+                for i in range(1, 11):
+                    if res is not None:
+                        break
+                    _LOGGER.info(f"Advising failed, retrying ({i}/10)")
+                    res = lib.advise_here(force=True)
+
+                if res is None:
+                    _LOGGER.error("Advise failed on server side, contact the maintainer")
+                    return False
+                _LOGGER.debug(f"{json.dumps(res)}")
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: f-string not needed.

File: kebechet/managers/thoth_advise/thoth_advise.py [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+
+                if res[1] is False:
+                    _LOGGER.info('Advise succeeded')
+                    self._write_advise(res)
+                    self._open_merge_request(branch_name, ["bot"], ["Pipfile.lock"])
```
Comment (fridex) [inline kebechet/managers/thoth_advise/thoth_advise.py]: Can these labels be parametrized (taken from the configuration file)?

File: kebechet/managers/thoth_provenance/thoth_provenance.py [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+            f"{checksum} - {len(error_info)}: Automated kebechet thoth-provenance Issue",
+            lambda: text_block,
+            labels=labels,
+        )
+
+    def run(self, labels: list):
+        """Run the provenance check bot."""
+        with cloned_repo(self.service_url, self.slug, depth=1) as repo:
+            self.repo = repo
+            if os.path.isfile("Pipfile") and os.path.isfile("Pipfile.lock"):
```
Comment (fridex) [inline kebechet/managers/thoth_provenance/thoth_provenance.py]: If not? I would expect an issue created on my repository, that bot cannot operate on it instead of silently skipping checks (which can be dangerous sometimes).

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 8 above)
Comment (KPostOffice) [reply kebechet/managers/thoth_advise/thoth_advise.py]: If the first advise fails it caches the failure so you have to force

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 12 above)
Comment (KPostOffice) [reply kebechet/managers/thoth_advise/thoth_advise.py]: Yeah, didn't realize this was the case

File: kebechet/managers/thoth_advise/thoth_advise.py (hunk 8 above)
Comment (fridex) [reply kebechet/managers/thoth_advise/thoth_advise.py]: Yes, but that's caused by the recent changes (bug). This is not something that should be happening and kebechet should handle.

**Comment by KPostOffice** [comment]
These managers are being split to reduce time where kebechet build is spent waiting
